crates/symbolicator-service/src/download/index.rs

```diff
@@ -0,0 +1,252 @@
+        writeln!(&mut cache_key, "segment: {i}").unwrap();
+        let cache_key = cache_key.build();
+        futures.push(cache.compute_memoized(request, cache_key));
+    }
+
+    let mut index = SymstoreIndex::default();
+    while let Some(result) = futures.next().await {
+        match result.into_contents() {
+            Ok(segment) => index.append(segment),
+            Err(e) => {
+                tracing::error!(
+                    error = &e as &dyn std::error::Error,
+                    "Failed to download index segment",
+                );
+            }
+        }
+    }
+
+    index.write(file)?;
+
+    Ok(())
+}
+
+impl CacheItemRequest for FetchSymstoreIndexSegment {
+    type Item = SymstoreIndex;
+
+    const VERSIONS: CacheVersions = SYMSTORE_INDEX_VERSIONS;
+
+    fn compute<'a>(&'a self, temp_file: &'a mut NamedTempFile) -> BoxFuture<'a, CacheContents> {
+        let downloader = Arc::clone(&self.downloader);
+        let source = Arc::clone(&self.source);
+
+        Box::pin(download_index_segment(
+            downloader,
+            source,
+            self.segment_id,
+            temp_file.as_file_mut(),
+        ))
+    }
+
+    fn load(&self, data: ByteView<'static>) -> CacheContents<Self::Item> {
+        let _result = data.hint(AccessPattern::Sequential);
+        let index = Self::Item::load(&data)?;
+        Ok(index)
+    }
+
+    fn weight(item: &Self::Item) -> u32 {
+        item.iter().map(|file| file.len() as u32).sum()
+    }
+}
+
+impl CacheItemRequest for FetchSymstoreIndex {
+    type Item = SymstoreIndex;
+
+    const VERSIONS: CacheVersions = SYMSTORE_INDEX_VERSIONS;
+
+    fn compute<'a>(&'a self, temp_file: &'a mut NamedTempFile) -> BoxFuture<'a, CacheContents> {
+        let downloader = Arc::clone(&self.downloader);
+        let source = Arc::clone(&self.source);
+
+        Box::pin(download_full_index(
+            self.segment_cache.clone(),
+            downloader,
+            source,
+            self.scope.clone(),
+            self.last_id,
+            temp_file.as_file_mut(),
+        ))
+    }
+
+    fn load(&self, data: ByteView<'static>) -> CacheContents<Self::Item> {
+        let _result = data.hint(AccessPattern::Sequential);
+        let index = Self::Item::load(&data)?;
+        Ok(index)
+    }
+
+    fn weight(item: &Self::Item) -> u32 {
+        item.iter().map(|file| file.len() as u32).sum()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SymstoreIndexService {
+    cache: Arc<Cacher<FetchSymstoreIndex>>,
+    segment_cache: Arc<Cacher<FetchSymstoreIndexSegment>>,
+    downloader: Arc<DownloadService>,
+    last_id_cache: moka::future::Cache<(Scope, SourceId), CacheContents<u32>>,
+}
+
+impl SymstoreIndexService {
+    pub fn new(
+        cache: Cache,
+        shared_cache: SharedCacheRef,
+        downloader: Arc<DownloadService>,
+    ) -> Self {
+        let last_id_cache = moka::future::Cache::builder()
+            .max_capacity(100)
```

Comment:
That essentially means we can store maximum `100` indices in memory?

Reply:
No, 100 last ids.

Reply:
100 last ids for 100 different indices, right?